### PR TITLE
ISSUE 64: Source pointers support

### PR DESCRIPTION
**Pull request by pavelpatrin**
https://github.com/jmattheis/goverter/issues/64

### CODE REVIEW

File: generator/generator.go [hunk 1]
```diff
@@ -408,21 +409,53 @@ func (g *generator) Build(
...
+		}
+
 		method := &methodDefinition{
 			ID:     name,
 			Name:   name,
 			Source: xtype.TypeOf(source.T),
 			Target: xtype.TypeOf(target.T),
-			Fields: map[string]*builder.FieldMapping{},
-			Flags:  g.flags,
+			Fields: methodFields,
```
Comment (pavelpatrin) [inline generator/generator.go:453]: Please pay attention on this. I'm worrying about side effects, because this method is picked up from the cache by signature, but signature is unaware of method flags and fields.

Without this, `govertor` fails on copying renamed or different case fields, because internal function is not aware of renames and case match.

With this change, `govertor` will handle fields renames in internal functions, but rules of these conversions will be taken from first method who spawned this function.

Only one solution I know is to add flags and fields to the method signature, but it is a huge change, I want to ask author's opinion first.

@jmattheis

File: generator/generator.go [hunk 2]
```diff
@@ -408,21 +409,53 @@ func (g *generator) Build(
...
+			Flags:  methodFlags,
 			Call:   jen.Id(xtype.ThisVar).Dot(name),
 		}
 
 		g.lookup[signature] = method
-		g.namer.Register(method.Name)
```
Comment (pavelpatrin) [inline generator/generator.go:425]: The `namer.Register()` is already called via `g.namer.Name(...)` several lines above.

File: generator/generator.go (hunk 1 above)
Comment (pavelpatrin) [reply generator/generator.go:453]: Or, as another way, just generate this function every time without adding it to cache and lookup for in in different methods. This will not add a lot of code to the result binary, but will significantly reduce generator complexity.

File: generator/generator.go [hunk 3]
```diff
@@ -408,21 +409,53 @@ func (g *generator) Build(
 		}
 	}
 
-	if !hasPointerStructMethod && ((source.Named && !source.Basic) || (target.Named && !target.Basic) || (source.Pointer && source.PointerInner.Named && !source.PointerInner.Basic)) {
+	isSourceNamed, isTargetNamed := source.Named && !source.Basic, target.Named && !target.Basic
+	isSourcePointerNamed := source.Pointer && source.PointerInner.Named && !source.PointerInner.Basic
+	isTargetPointerNamed := target.Pointer && target.PointerInner.Named && !target.PointerInner.Basic
+	if !hasPointerStructMethod && (isSourceNamed || isTargetNamed || isSourcePointerNamed || isTargetPointerNamed) {
```
Comment (jmattheis) [inline generator/generator.go:415]: Why this change? I reverted the 4 lines above and no test failed.

File: scenario/auto_map_pointer_source.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    func (c *ConverterImpl) Convert(source execution.Person) execution.FlatPerson {
+    	var exampleFlatPerson execution.FlatPerson
+    	exampleFlatPerson.Name = source.Name
+    	exampleFlatPerson.Age = source.Age
+    	var pString *string
+    	if source.Address != nil {
+    		pString = &source.Address.Street
+    	}
+    	xstring := *pString
```
Comment (jmattheis) [inline scenario/auto_map_pointer_source.yml:41]: This code will produce a
```
panic: runtime error: invalid memory address or nil pointer dereference
```
goverter must never panic, the code could be written like this to prevent this.
```go
    func (c *ConverterImpl) Convert(source execution.Person) execution.FlatPerson {
    	var exampleFlatPerson execution.FlatPerson
    	exampleFlatPerson.Name = source.Name
    	exampleFlatPerson.Age = source.Age
    	var xstring string
    	if source.Address != nil {
    		xstring = (*source.Address).Street
    	}
    	exampleFlatPerson.Street = xstring
    	return exampleFlatPerson
    }
```

File: builder/pointer.go [hunk 5]
```diff
@@ -52,6 +52,43 @@ func (*Pointer) Build(gen Generator, ctx *MethodContext, sourceID *xtype.JenID,
...
+// SourcePointer handles type were only the source is a pointer.
+type SourcePointer struct{}
+
+// Matches returns true, if the builder can create handle the given types.
+func (*SourcePointer) Matches(source, target *xtype.Type) bool {
+	return source.Pointer && !target.Pointer
+}
+
+// Build creates conversion source code for the given source and target type.
+func (*SourcePointer) Build(gen Generator, ctx *MethodContext, sourceID *xtype.JenID, source, target *xtype.Type) ([]jen.Code, *xtype.JenID, *Error) {
```
Comment (jmattheis) [inline builder/pointer.go:64]: This behavior should not be enabled by default, it should be behind a converter comment. See https://github.com/jmattheis/goverter/issues/11 for my reasoning.

File: generator/generator.go (hunk 1 above)
Comment (jmattheis) [reply generator/generator.go:453]: Both the method flags and method fields should _not_ be propagated to other generator methods, as this adds possible clashes with other methods. Example:
```
// goverter:converter
type Converter interface {
    // goverter:matchIgnoreCase
	Convert1(Input1) Output
	Convert2(Input2) Output
}

type Input1 struct { Inner InputInner }
type Input2 struct { Inner InputInner }
type Output struct { Inner OutputInner }
type InputInner struct { Age int }
type OutputInner struct { AGE int }
```
Now it's unclear if the internal generated method converting from InputInner to OutputInner should have matchIgnoreCase enabled or not, therefore the flags should not propagate, and the user should add the flag like this:
```
// goverter:converter
type Converter interface {
	Convert1(Input1) Output
	Convert2(Input2) Output

	// goverter:matchIgnoreCase
	ConvertInner(InputInner) OutputInner
}
```
this is also described here: https://goverter.jmattheis.de/#/conversion/configure-nested

**Review by jmattheis — CHANGES_REQUESTED** [review]

File: generator/generator.go (hunk 1 above)
Comment (jmattheis) [reply generator/generator.go:453]: Both the method flags and method fields should not be propagated to other generator methods, as this adds possible clashes with other methods. Example:
```go
// goverter:converter
type Converter interface {
    // goverter:matchIgnoreCase
	Convert1(Input1) Output
	Convert2(Input2) Output
}

type Input1 struct { Inner InputInner }
type Input2 struct { Inner InputInner }
type Output struct { Inner OutputInner }
type InputInner struct { Age int }
type OutputInner struct { AGE int }
```
Now it's unclear if the internal generated method converting from InputInner to OutputInner should have matchIgnoreCase enabled or not, therefore the flags should not propagate, and the user should add the flag like this:
```go
// goverter:converter
type Converter interface {
	Convert1(Input1) Output
	Convert2(Input2) Output

	// goverter:matchIgnoreCase
	ConvertInner(InputInner) OutputInner
}
```
this is also described here: https://goverter.jmattheis.de/#/conversion/configure-nested

Comment (pavelpatrin) [reply generator/generator.go:453]: Thank you for the clarification. That means I was wrong with this.

File: generator/generator.go (hunk 3 above)
Comment (pavelpatrin) [reply generator/generator.go:415]: Let me think more about it.

File: scenario/auto_map_pointer_source.yml (hunk 4 above)
Comment (pavelpatrin) [reply scenario/auto_map_pointer_source.yml:41]: Agree.

File: builder/pointer.go (hunk 5 above)
Comment (pavelpatrin) [reply builder/pointer.go:64]: Thank your for giving understanding why it is important. Let me improve this.

File: generator/generator.go (hunk 1 above)
Comment (pavelpatrin) [reply generator/generator.go]: Got rid of this.

File: scenario/auto_map_pointer_source.yml (hunk 4 above)
Comment (pavelpatrin) [reply scenario/auto_map_pointer_source.yml]: Should be fixed.

File: builder/pointer.go (hunk 5 above)
Comment (pavelpatrin) [reply builder/pointer.go:75]: Added machinery to check status of the every converter by the method context. All converters but the `SourcePointer` are enabled by default. SourcePointer checks corresponding method flag.

File: generator/generator.go (hunk 3 above)
Comment (pavelpatrin) [reply generator/generator.go]: Got rid of this. Checked by tests.

File: scenario/slice_source_pointer.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
+input:
```
Comment (pavelpatrin) [inline scenario/slice_source_pointer.yml:1]: I'm not sure about test scenario file names, please make me on the right way.

File: scenario/slice_source_pointer.yml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    package generated
+
+    import execution "github.com/jmattheis/goverter/execution"
+
+    type ConverterImpl struct{}
+
+    func (c *ConverterImpl) ConvertHouse(source execution.DBHouseNames) execution.APIHouseNames {
+    	var slices_arraysAPIHouseNames execution.APIHouseNames
+    	var stringList []string
+    	if (*source.Names) != nil {
```
Comment (pavelpatrin) [inline scenario/slice_source_pointer.yml:30]: Additional parentheses there are looking redundant, but they are required for `(*source.Names)[i]` syntax. Looks like it is by-design.

File: scenario/auto_map_pointer_source.yml (hunk 4 above)
Comment (pavelpatrin) [reply scenario/auto_map_pointer_source.yml]: I found one more problem with null pointers. Please don't merge this right now.

Comment (pavelpatrin) [reply scenario/auto_map_pointer_source.yml]: Fixed in https://github.com/jmattheis/goverter/pull/65/commits/74698915f955025e99b39964d65b56b3ef43a42a.

Now implementation of if-block is looking really same to the `(*Pointer) Build` by the code layout.

**Review by jmattheis — APPROVED** [review]
I've made some changes:
* Move enabled into matches, the methods are basically the same but with different parameters
* Add more tests
* Rename goverter flag, it's now `goverter:useZeroValueOnPointerInconsistency`, it's pretty long but I couldn't come up with something better.
* Add docs
* Remove the `if id.Variable` from builder/pointer.go, both branches had the same code.
* Remove the cli flag; I dislike them, and you can already globally enable this feature per converter interface.
